subworkflows/local/preparegenome.nf

```diff
@@ -0,0 +1,155 @@
+
+include {
+    GUNZIP as GUNZIP_FASTA
+    GUNZIP as GUNZIP_GTF
+    GUNZIP as GUNZIP_GFF
+    GUNZIP as GUNZIP_GENE_BED
+    GUNZIP as GUNZIP_ADDITIONAL_FASTA } from '../../modules/nf-core/modules/gunzip/main'               addParams( options: params.options.gunzip        )
+include { GTF2BED                     } from '../../modules/local/gtf2bed'                             addParams( options: params.options.gtf2bed       )
+include { CHROMSIZES                  } from '../../modules/local/genome/chromsizes'                   addParams( options: params.options.chromsizes    )
+include { GENOME_FILTER               } from '../../modules/local/genome/filter'                       addParams( options: params.options.genomefilter  )
+include { COOLER_DIGEST               } from '../../modules/nf-core/modules/cooler/digest/main'        addParams( options: params.options.digest_genome )
+include { GFFREAD                     } from '../../modules/nf-core/modules/gffread/main'              addParams( options: params.options.gffread       )
+include { BWA_INDEX                   } from '../../modules/nf-core/modules/bwa/index/main'            addParams( options: params.options.bwa_index     )
+
+workflow PREPARE_GENOME {
+    main:
+
+    /*
+     * Uncompress genome fasta file if required
+     */
+    if (params.fasta.endsWith('.gz')) {
+        ch_fasta = GUNZIP_FASTA ( params.fasta ).gunzip
+    } else {
+        ch_fasta = file(params.fasta)
+    }
+
+    /*
+     * Uncompress GTF annotation file or create from GFF3 if required
+     */
+    ch_version = Channel.empty()
+    ch_gtf = Channel.empty()
+    if (params.gtf) {
+        if (params.gtf.endsWith('.gz')) {
+            ch_gtf = GUNZIP_GTF ( params.gtf ).gunzip
+        } else {
+            ch_gtf = file(params.gtf)
+        }
+    } else if (params.gff) {
+        if (params.gff.endsWith('.gz')) {
+            ch_gff = GUNZIP_GFF ( params.gff ).gunzip
+        } else {
+            ch_gff = file(params.gff)
+        }
+        ch_gtf = GFFREAD ( ch_gff ).gtf
+        ch_version = ch_version.mix(GFFREAD.out.versions.ifEmpty(null))
+    }
+
+    /*
+     * Uncompress gene BED annotation file or create from GTF if required
+     */
+    ch_gene_bed = Channel.empty()
+    if (params.gene_bed) {
+        if (params.gene_bed.endsWith('.gz')) {
+            ch_gene_bed = GUNZIP_GENE_BED ( params.gene_bed ).gunzip
+        } else {
+            ch_gene_bed = file(params.gene_bed)
+        }
+    } else {
+        ch_gene_bed = GTF2BED ( ch_gtf ).bed
+        ch_version = ch_version.mix(GTF2BED.out.versions.ifEmpty(null))
+    }
+
+    /*
+     * Create chromosome sizes file
+     */
+    ch_chrom_sizes = CHROMSIZES ( ch_fasta ).sizes
+
+    /*
+     * Calculate effective genome sizes
+     */
+    gs = ch_fasta.size() * 0.78
+    if (params.macs_gsize) {
+        gs = params.macs_gsize
+    } else {
+        //genome size remove all N then * 78%
+        gs = 0.0
+        ch_fasta.withReader{
```

Comment:
Again here, channels do not have a `withReader` operator. This only works on files.

Reply:
Sorry. I forget to fix this issue.